### ISSUE TITLE
Core.RemoveObjectFromList performance

### ISSUE THREAD

**Issue by juflo** [NetOfficeFw/NetOffice#221]
Hi

When an NetOffice COMObject is disposed, its children are also disposed : the method Core.RemoveObjectFromList is called.
The problem is the _globalObjectList in Core is of type List<COMObject>, which is of complexity O(n²) when calling list.Remove(child) : it's really slow.
It should be possible to use a HashSet instead to achieve far better performance.

**Comment by GuoHuiChen** [comment]
Yes , I always dispose the COM Objects as soon as possible, use using(var a = ....){}

**Comment by jozefizso** [comment]
## Analysis

I've analyzed this performance issue and can confirm the problem. Here's a detailed breakdown:

### Current Implementation

The `_globalObjectList` is declared as a `List<ICOMObject>` in `Source/NetOffice/Core.cs:96`:

```csharp
private List<ICOMObject> _globalObjectList = new List<ICOMObject>();
```

When `RemoveObjectFromList` is called (Core.cs:1346-1367), it uses `List<T>.Remove()`:

```csharp
internal void RemoveObjectFromList(ICOMObject proxy, IEnumerable<ICOMObject> ownerPath)
{         
    try
    {
        bool removed = false;
        lock (_globalObjectList)
        {
            removed = _globalObjectList.Remove(proxy);  // O(n) operation
            
            if (HasProxyRemovedRecipients)
            {
                RaiseProxyRemoved(ownerPath, proxy);
            }
        }
        if (removed)
            RaiseProxyCountChanged(_globalObjectList.Count);
    }
    catch (Exception throwedException)
    {
        Console.WriteException(throwedException);
    }
}
```

### Performance Problem

**Time Complexity:**
- `List<T>.Remove()` is **O(n)** - it must search the entire list sequentially and then shift elements
- When disposing an object with many children, each child triggers a removal
- This creates **O(n²) complexity** for batch disposals
- The problem worsens as the object graph grows

**Real-World Impact:**
Users who follow best practices (using `using` statements and proper disposal) are **penalized** with worse performance. The more diligent you are about cleanup, the slower it becomes.

### Call Sites

`RemoveObjectFromList` is called from 4 locations:
1. **COMObject.cs:656** - During object disposal (main bottleneck)
2. **COMDynamicObject.cs:472** - During dynamic object disposal  
3. **COMObject.cs:168, 211** - When replacing objects
4. **Core.cs:1289** - During instance replacement

### Recommended Solution

Replace `List<ICOMObject>` with `HashSet<ICOMObject>`:

**Benefits:**
- `HashSet<T>.Remove()` is **O(1)** vs O(n) for `List<T>`
- Reduces disposal complexity from **O(n²) to O(n)**
- Minimal code changes required
- No breaking changes to public API

**Implementation Considerations:**

1. **Index Access**: `DisposeAllCOMProxies()` (line 1310) accesses by index: `_globalObjectList[_globalObjectList.Count - 1]`. This would need to be changed to iterate or convert to array.

2. **Thread Safety**: Current locking strategy works fine with HashSet

3. **Count & Enumeration**: Both `List` and `HashSet` support `.Count` and enumeration, so `ProxyCount` property and `GetRootInstances()` need no changes

### Alternative Approach

If ordering is important for disposal (disposing in reverse creation order), consider `LinkedList<ICOMObject>` with a `Dictionary<ICOMObject, LinkedListNode<ICOMObject>>` lookup:
- Maintains insertion order
- O(1) removal with dictionary lookup
- More memory overhead but still much better than current O(n²) behavior
